File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/WhereIsCommand.cs
using Atlasd.Daemon;
using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class WhereIsCommand : ChatCommand
    {
        public WhereIsCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            var t = Arguments.Count == 0 ? "" : Arguments[0]; // target
            string r; // reply

            // quick check
            if (t.ToLower() == context.GameState.OnlineName.ToLower())
            {
                new WhoAmICommand(RawBuffer, Arguments).Invoke(context);
                return;
            }

            if (!Battlenet.Common.GetClientByOnlineName(t, out var target) || target == null)
            {
                r = Resources.UserNotLoggedOn;
                foreach (var line in r.Split(Battlenet.Common.NewLine))
                    new ChatEvent(ChatEvent.EventIds.EID_ERROR, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
                return;
            }

            if (target == context.GameState)
            {
                new WhoAmICommand(RawBuffer, Arguments).Invoke(context);
                return;
            }

            var ch = target.ActiveChannel;
            var g = target.GameAd;

            if (g != null) // TODO: Consider private games and friendship.
                r = Resources.UserIsUsingGameInTheGame;
            else if (ch != null)
                r = Resources.UserIsUsingGameInTheChannel;
            else
                r = Resources.UserIsUsingGameInRealm;

            if (target.Away != null)
                r += Battlenet.Common.NewLine + Resources.AwayCommandStatus.Replace("{awayMessage}", target.Away);

            var targetEnv = new Dictionary<string, string>()
            {
                { "accountName", target.Username },
                { "channel", target.ActiveChannel == null ? "(null)" : target.ActiveChannel.Name },
                { "game", Product.ProductName(target.Product, true) },
                { "gameAd", g == null ? "(null)" : Encoding.UTF8.GetString(g.Name) },
                { "host", Settings.GetString(new string[] { "battlenet", "realm", "host" }, "(null)") },
                { "localTime", target.LocalTime.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ") },
                { "name", Channel.RenderOnlineName(context.GameState, target) },
                { "onlineName", Channel.RenderOnlineName(context.GameState, target) },
                { "realm", Settings.GetString(new string[] { "battlenet", "realm", "name" }, Resources.Battlenet) },
                { "realmTime", DateTime.Now.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ") },
                { "realmTimezone", $"UTC{DateTime.Now:zzz}" },
                { "user", Channel.RenderOnlineName(context.GameState, target) },
                { "username", Channel.RenderOnlineName(context.GameState, target) },
                { "userName", Channel.RenderOnlineName(context.GameState, target) },
            };
            var env = targetEnv.Concat(context.Environment);

            foreach (var kv in env) r = r.Replace("{" + kv.Key + "}", kv.Value);
            new ChatEvent(ChatEvent.EventIds.EID_INFO, Channel.RenderChannelFlags(context.GameState, target), target.Ping, Channel.RenderOnlineName(context.GameState, target), r).WriteTo(context.GameState.Client);
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/WhereIsCommand.vb
Imports AtlasV.Localization
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class WhereIsCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim t = If(Arguments.Count = 0, "", Arguments(0))
            Dim r As String

            If t.ToLower() = varContext.GameState.OnlineName.ToLower() Then
                Call New WhoAmICommand(RawBuffer, Arguments).Invoke(varContext)
                Return
            End If

            Dim target As GameState = Nothing

            If Not Battlenet.Common.GetClientByOnlineName(t, target) OrElse target Is Nothing Then
                r = Resources.UserNotLoggedOn

                For Each line In r.Split(Battlenet.Common.NewLine)
                    Call New ChatEvent(ChatEvent.EventIds.EID_ERROR, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
                Next

                Return
            End If

            If Object.Equals(target, varContext.GameState) = True Then
                Call New WhoAmICommand(RawBuffer, Arguments).Invoke(varContext)
                Return
            End If

            Dim ch = target.ActiveChannel
            Dim str = If(ch Is Nothing, Resources.UserIsUsingGameInRealm, Resources.UserIsUsingGameInTheChannel)

            If target.Away IsNot Nothing Then
                str += Battlenet.Common.NewLine + Resources.AwayCommandStatus.Replace("{awayMessage}", target.Away)
            End If

            Dim targetEnv = New Dictionary(Of String, String)() From {
                {"accountName", target.Username},
                {"channel", If(target.ActiveChannel Is Nothing, "(null)", target.ActiveChannel.Name)},
                {"game", Product.ProductName(target.Product, True)},
                {"host", "BNETDocs"},
                {"localTime", target.LocalTime.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ")},
                {"name", Channel.RenderOnlineName(varContext.GameState, target)},
                {"onlineName", Channel.RenderOnlineName(varContext.GameState, target)},
                {"realm", "BNETDocs"},
                {"realmTime", DateTime.Now.ToString(Common.HumanDateTimeFormat).Replace(" 0", "  ")},
                {"realmTimezone", $"UTC{DateTime.Now}"},
                {"user", Channel.RenderOnlineName(varContext.GameState, target)},
                {"username", Channel.RenderOnlineName(varContext.GameState, target)},
                {"userName", Channel.RenderOnlineName(varContext.GameState, target)}
            }
            Dim env = targetEnv.Concat(varContext.Environment)

            For Each kv In env
                str = str.Replace("{" & kv.Key & "}", kv.Value)
            Next

            Call New ChatEvent(ChatEvent.EventIds.EID_INFO, Channel.RenderChannelFlags(varContext.GameState, target), target.Ping, Channel.RenderOnlineName(varContext.GameState, target), str).WriteTo(varContext.GameState.Client)
        End Sub
    End Class
End Namespace
